research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -1,255 +1,145 @@
-            self.reset()
-            raise StopIteration
-        else:
-            self.current_step += 1
-            batch = next(self.nnunet_dataloader)  # This returns a dictionary
-            # Note: When deep supervision is on, target is a list of segmentations at various scales
-            # nnUNet has a wrapper for loss functions to enable deep supervision
-            inputs: torch.Tensor = batch["data"]
-            if isinstance(batch["target"], list):
-                target_dict = convert_ds_list_to_dict(batch["target"], self.num_spatial_dims)
-                return inputs, target_dict
-            elif isinstance(batch["target"], torch.Tensor):
-                target_tensor: torch.Tensor = batch["target"]
-                return inputs, target_tensor
-            else:
-                raise TypeError(
-                    "Was expecting the target generated by the nnunet dataloader to be a list or a torch.Tensor"
-                )
-
-    def __len__(self) -> int:
-        # Should return num_samples // batch_size
-        # nnUNetDataloaders are 'infinite' meaning they randomly sample batches
-        # from the dataset. This makes the distinction between
-        # num_samples // batch_size and ceil(num_samples/batch_size)
-        # meaningless. We will abritrarily use the later
-        num_samples = len(self.dataset)
-        batch_size = self.nnunet_dataloader.generator.batch_size
-        return int(ceil(num_samples / batch_size))
-
-    def reset(self) -> None:
-        self.current_step = 0
-
-    def __iter__(self) -> DataLoader:  # type: ignore
-        # mypy gets angry that the return type
-        return self
-
-
-class Module2LossWrapper(_Loss):
-    """Converts a nn.Module subclass to a _Loss subclass"""
-
-    def __init__(self, loss: nn.Module, **kwargs: Any) -> None:
-        super().__init__(**kwargs)
-        self.loss = loss
-
-    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
-        return self.loss(pred, target)
+# Get the default signal handlers used by python before flwr overrides them
+# We need these because the nnunet dataloaders spawn child processes
+# and flwr throws errors when those processes end. So we set the signal handlers
+# for the child processes to the python defaults to avoid this
+ORIGINAL_SIGINT_HANDLER = signal.getsignal(signal.SIGINT)
+ORIGINAL_SIGTERM_HANDLER = signal.getsignal(signal.SIGTERM)
 
 
 class nnUNetClient(BasicClient):
     def __init__(
         self,
+        data_path: Path,
+        device: torch.device,
         dataset_id: int,
-        data_identifier: Optional[str],
-        plans_identifier: Optional[str],
-        always_preprocess: bool = True,
-        **kwargs: Any,
+        fold: Union[int, str],
+        data_identifier: Optional[str] = None,
+        plans_identifier: Optional[str] = None,
+        always_preprocess: bool = False,
+        metrics: Sequence[Metric] = [],
```

Comment:
Thanks for expanding out the arguments 🙂. One thing to note is that defaults for mutable objects in python can produce unexpected behavior ([see here](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments)). To get default empty list behavior without this behavior, the typical workaround is an metrics: Optional[Sequence[Metric]] = None and in the __init__ function doing
```
self.metrics = metrics if metrics else []
```